clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/OperationsService.java

```diff
@@ -0,0 +1,252 @@
+import com.netflix.spinnaker.clouddriver.security.ProviderVersion;
+import com.netflix.spinnaker.kork.exceptions.SystemException;
+import com.netflix.spinnaker.security.AuthenticatedRequest;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.Data;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.ResolvableType;
+import org.springframework.validation.Errors;
+
+@Slf4j
+public class OperationsService {
+
+  private final Splitter COMMA_SPLITTER = Splitter.on(",");
+
+  private final AtomicOperationsRegistry atomicOperationsRegistry;
+  private final DescriptionAuthorizer descriptionAuthorizer;
+  private final Collection<AllowedAccountsValidator> allowedAccountValidators;
+  private final List<AtomicOperationDescriptionPreProcessor>
+      atomicOperationDescriptionPreProcessors;
+  private final AccountCredentialsRepository accountCredentialsRepository;
+  private final Registry registry;
+  private final ObjectMapper objectMapper;
+
+  private final Id validationErrorsCounterId;
+
+  public OperationsService(
+      AtomicOperationsRegistry atomicOperationsRegistry,
+      DescriptionAuthorizer descriptionAuthorizer,
+      Optional<Collection<AllowedAccountsValidator>> allowedAccountValidators,
+      Optional<List<AtomicOperationDescriptionPreProcessor>>
+          atomicOperationDescriptionPreProcessors,
+      AccountCredentialsRepository accountCredentialsRepository,
+      Registry registry,
+      ObjectMapper objectMapper) {
+    this.atomicOperationsRegistry = atomicOperationsRegistry;
+    this.descriptionAuthorizer = descriptionAuthorizer;
+    this.allowedAccountValidators = allowedAccountValidators.orElse(Collections.emptyList());
+    this.atomicOperationDescriptionPreProcessors =
+        atomicOperationDescriptionPreProcessors.orElse(Collections.emptyList());
+    this.accountCredentialsRepository = accountCredentialsRepository;
+    this.registry = registry;
+    this.objectMapper = objectMapper;
+
+    validationErrorsCounterId = registry.createId("validationErrors");
+  }
+
+  @Nonnull
+  public List<AtomicOperation> collectAtomicOperations(@Nonnull List<Map<String, Map>> inputs) {
+    return collectAtomicOperations(null, inputs);
+  }
+
+  @Nonnull
+  public List<AtomicOperation> collectAtomicOperations(
+      @Nullable String cloudProvider, @Nonnull List<Map<String, Map>> inputs) {
+    List<AtomicOperationBindingResult> results = convert(cloudProvider, inputs);
+
+    List<AtomicOperation> atomicOperations = new ArrayList<>();
+    results.forEach(
+        bindingResult -> {
+          if (bindingResult.errors.hasErrors()) {
+            throw new DescriptionValidationException(bindingResult.errors);
```

Comment:
without being a language lawyer I'm not quite sure what the expected behavior is here, throwing exceptions in a lambda in a `forEach`? Are we expecting an exception for each result that has an error? Are we getting out on the first error?

Reply:
We're getting out on the first error. Changing this logic would alter backwards compatibility. Again, this was originally in a controller, so it was just a lift/shift.